### PR TITLE
llvm: Don’t copy “prefix/bin” into “prefix” during install

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

**Comment by alalazo** [comment]
I added @trws as a reviewer as I think he may be interested in this changeset.

**Comment by trws** [comment]
Why do you want to make this change @eschnett? The `cp` was added to copy the tools and components not installed by `make install` but which are required for certain projects.  Has LLVM started installing all built binaries?

**Comment by eschnett** [comment]
Copying tools into the prefix directory clutters it. Usually, tools are installed into `bin` or similar. (Also, using `cp` this way instead of `install_tree` seemed like a left-over debug statement.)

**Comment by trws** [comment]
Fair enough put then in bin then, which is clearly where they should be as you point out.

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -374,5 +374,3 @@ def install(self, spec, prefix):
             cmake(*cmake_args)
             make()
             make("install")
-            cp = which('cp')
-            cp('-a', 'bin/', prefix)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/llvm/package.py]: Please change this to:
```python
install_tree('bin', prefix.bin)
```

Comment (eschnett) [inline var/spack/repos/builtin/packages/llvm/package.py]: Done.

**Comment by eschnett** [comment]
Unfortunately, that didn't work, as `install_tree` doesn't overwrite. A `shutil.rmtree` or similar is missing (pushed onto branch).

**Comment by adamjstewart** [comment]
You'll have to create a new branch + PR. I don't believe there's a way to merge additional commits after a PR has been merged.

Were there already files in prefix.bin? We probably want to keep them.

**Comment by eschnett** [comment]
I'll open a new PR.

Yes, `make install` already installs files into `prefix.bin`. I assume these are a subset of those that are created in stage/`bin`.

Are you sure we want to install all the files from the build directory? Some of the binaries have names like `count`, `not`, `update_check.py`. These are quite generic, and are likely to conflict with similar names from other random packages. Should we install them elsewhere, e.g. `prefix.libexec`/llvm/bin?

**Comment by adamjstewart** [comment]
@trws? If `make install` puts things in bin already, I'm inclined to think that the LLVM developers don't want anything else installed.

**Comment by trws** [comment]
It may be fair to use a different path, but I've run into more than one component that either I needed or another package used that isn't installed by default. The clang-query utility comes to mind, and some of the check scripts used by out of tree clang tool builds.


---
Sent from Boxer | http://getboxer.com<http://bxr.io/PBI3C>

On March 21, 2017 at 10:26:29 AM EDT, Adam J. Stewart <notifications@github.com> wrote:

@trws<https://github.com/trws>? If make install puts things in bin already, I'm inclined to think that the LLVM developers don't want anything else installed.

**Comment by eschnett** [comment]
See #3516 .

**Pull request merged**